src/probnum/diffeq/odefiltsmooth/kalman_odesolution.py

```diff
@@ -79,66 +71,109 @@ class KalmanODESolution(ODESolution):
-
-    @cached_property
-    def dy(self) -> pnrv_list._RandomVariableList:
-        dy_rvs = [
-            _project_rv(self.proj_to_dy, rv) for rv in self.kalman_posterior.state_rvs
-        ]
-        return pnrv_list._RandomVariableList(dy_rvs)
-
-    def __call__(
-        self, t: float
-    ) -> typing.Union[pnrv.RandomVariable, pnrv_list._RandomVariableList]:
-        out_rv = self.kalman_posterior(t)
-
-        if np.isscalar(t):
-            return _project_rv(self.proj_to_y, out_rv)
-
-        return pnrv_list._RandomVariableList(
-            [_project_rv(self.proj_to_y, rv) for rv in out_rv]
+        states = _randomvariablelist._RandomVariableList(
+            [_project_rv(self.proj_to_y, rv) for rv in self.kalman_posterior.states]
+        )
+        derivatives = _randomvariablelist._RandomVariableList(
+            [_project_rv(self.proj_to_dy, rv) for rv in self.kalman_posterior.states]
         )
+        super().__init__(
+            locations=kalman_posterior.locations, states=states, derivatives=derivatives
+        )
+
+    def interpolate(self, t: FloatArgType) -> random_variables.RandomVariable:
+        out_rv = self.kalman_posterior.interpolate(t)
+        return _project_rv(self.proj_to_y, out_rv)
 
     def sample(
         self,
-        t: typing.Optional[float] = None,
-        size: typing.Optional[probnum.type.ShapeArgType] = (),
+        t: Optional[DenseOutputLocationArgType] = None,
+        size: Optional[ShapeArgType] = (),
+        random_state: Optional[RandomStateArgType] = None,
     ) -> np.ndarray:
-        """Sample from the Gaussian filtering ODE solution by sampling from the Gauss-
-        Markov posterior."""
-        size = probnum.utils.as_shape(size)
 
-        # implement only single samples, rest via recursion
-        # We cannot 'steal' the recursion from self.kalman_posterior.sample,
-        # because we need to project the respective states out of each sample.
-        if size != ():
-            return np.array([self.sample(t=t, size=size[1:]) for _ in range(size[0])])
+        # Include the final point if a specific grid is demanded
+        # and the rightmost point is left of the rightmost data point.
+        # If this is not done, the samples are not from the full posterior.
+        if t is None:
+            sampling_locs = self.locations
+            remove_final_point = False
+        elif t[-1] >= self.locations[-1]:
+            sampling_locs = t
+            remove_final_point = False
+        else:
+            sampling_locs = np.hstack((t, self.locations[-1]))
+            remove_final_point = True
+
+        # Infer desired size of the base measure realizations and create them
+        size = utils.as_shape(size)
+        single_rv_shape = self.kalman_posterior.states[0].shape
+        base_measure_realizations = stats.norm.rvs(
+            size=(size + sampling_locs.shape + single_rv_shape),
+            random_state=random_state,
+        )
+
```

Comment:
That is a good point. The problem is here how to slice the correct dimensions in a way that works with all kinds of samplign shapes (which I wasnt able to do, so I mirrored the KalmanPosterior logic in KalmanODESolution). This could be done with np.einsum I suppose, let me try some things and get back to rou.

Reply:
Ignore the previous message. Here is the proper answer:

This is indeed a potentially controversial point.
We have to decide between (a) letting `KalmanODESolution` have a `transform_base_measure_realizations` which `sample()` points to, and (b) letting `KalmanODESolution.sample()` point to `KalmanPosterior.sample()` (which itself points to a `KalmanPosterior.transform_base_measure...`. 

In the spirit that `sample()` is something that "sits on top of `transform_base...`", I opted for (a). In other words, `KalmanODESolution.sample()` delegates the implementation to `KalmanPosterior` via the `transform_base_measure_realizations()` route. 

On a side note, you're right in that `KalmanODESolution.transform_base_measure_realizations` could probably do without the recursion itself. Let me see what I can do in this regard.

Reply:
> On a side note, you're right in that KalmanODESolution.transform_base_measure_realizations could probably do without the recursion itself. Let me see what I can do in this regard.

This is taken care of.